kubernetes/scheduler/controller.go

```diff
@@ -0,0 +1,668 @@
+		case pod.Status.InitContainerStatuses[0].State.Running != nil:
+			c.Debug("init container running", "pod", pod.ObjectMeta.Name)
+			c.Debug("calling CopyTar", "pod", pod.ObjectMeta.Name, "files", c.opts.Files)
+			err := c.libclient.CopyTar(c.opts.Files, pod)
+			// If this errors the controller will not die. It will just be
+			// logged.
+			if err != nil {
+				c.sendErrChan(fmt.Sprintf("CopyTar for pod %s failed: %s", pod.ObjectMeta.Name, err))
+				return err
+			}
+		default:
+		}
+	}
+	// Error handling
+	if pod.Status.Phase == corev1.PodSucceeded {
+		c.Debug("pod exited succesfully, notifying", "pod", pod.ObjectMeta.Name)
+		//jsonObj, _ := json.Marshal(pod) c.Info(fmt.Sprintf("Contents of pod
+		//%s", jsonObj))
+		// Get the pod's errChan, return nil signifying that the pod
+		// (runner) exited succesfully.
+		result, ok := c.podAliveMap.Load(pod.ObjectMeta.UID)
+		if ok {
+			go func() {
+				req := result.(*PodAlive)
+				if !req.Done {
+					req.Done = true
+					c.Info("sending nil on error chan", "pod", pod.ObjectMeta.Name)
+					req.ErrChan <- nil
+					close(req.ErrChan)
+				}
+			}()
+		} else {
+			c.Error("could not find return error channel", "pod", pod.ObjectMeta.Name)
+			return fmt.Errorf("Could not find return error channel for pod %s", pod.ObjectMeta.Name)
+		}
+	}
+	// Entire pod has failed for some reason.
+	if pod.Status.Phase == corev1.PodFailed {
+		result, ok := c.podAliveMap.Load(pod.ObjectMeta.UID)
+		if ok {
+			go func() {
+				req := result.(*PodAlive)
+				if !req.Done {
+					req.Done = true
+					c.Error("sending error on error chan", "pod", pod.ObjectMeta.Name, "error", "pod failed")
+					req.ErrChan <- fmt.Errorf("Pod %s failed", pod.ObjectMeta.Name)
+					close(req.ErrChan)
+				}
+			}()
+		} else {
+			c.Error("could not find return error channel", "pod", pod.ObjectMeta.Name)
+			return fmt.Errorf("Could not find return error channel for pod %s", pod.ObjectMeta.Name)
+		}
+		// Get logs
+		logs, err := c.libclient.GetLog(pod, 25)
+		if err != nil {
+			c.Error("failed to get logs", "pod", pod.ObjectMeta.Name, "err", err)
+		}
+		// Callback to user, trying to give an informative error.
+		c.Error("Pod failed", "pod", pod.ObjectMeta.Name, "message", pod.Status.Message, "reason", pod.Status.Reason)
+		c.sendErrChan(fmt.Sprintf("Pod %s failed. Reason: %s, Message: %s\n Logs: %s", pod.ObjectMeta.Name, pod.Status.Message, pod.Status.Reason, logs))
+		c.sendBadServer(&cloud.Server{
+			Name: pod.ObjectMeta.Name,
+		})
+	}
+	// Pod is pending (cluster has not got enough capacity.)
+	if pod.Status.Phase == corev1.PodPending {
+		// If the request on the container is currently not possible, send a
+		// callback to the user
+		for _, condition := range pod.Status.Conditions {
+			if condition.Type == corev1.PodScheduled && condition.Reason == "Unschedulable" {
+				c.sendErrChan(fmt.Sprintf("Pod %s pending, reason: %s", pod.ObjectMeta.Name, condition.Message))
```

Comment:
Did you think about excluding pod name under certain conditions? Even if that means checking for certain strings in .Message?

Reply:
I'm thinking about writing a function to go from k8s resources -> ints (or approximately) so I can do the modifications to canCount(). If I can do that I should be able to use the same function here to not return a pod name but the resources requested and group messages like that?